core/API/Request.php

```diff
@@ -263,27 +292,53 @@ public static function reloadAuthUsingTokenAuth($request = null)
     {
         // if a token_auth is specified in the API request, we load the right permissions
         $token_auth = Common::getRequestVar('token_auth', '', 'string', $request);
-        $access = Access::getInstance();
-
-        if ($token_auth && $token_auth !== $access->getTokenAuth()) {
-
-            /**
-             * Triggered when authenticating an API request, but only if the **token_auth**
-             * query parameter is found in the request.
-             *
-             * Plugins that provide authentication capabilities should subscribe to this event
-             * and make sure the global authentication object (the object returned by `StaticContainer::get('Piwik\Auth')`)
-             * is setup to use `$token_auth` when its `authenticate()` method is executed.
-             *
-             * @param string $token_auth The value of the **token_auth** query parameter.
-             */
-            Piwik::postEvent('API.Request.authenticate', array($token_auth));
-            Access::getInstance()->reloadAccess();
-            SettingsServer::raiseMemoryLimitIfNecessary();
+
+        if (self::shouldReloadAuthUsingTokenAuth($request)) {
+            self::doReloadAuthUsingTokenAuth($token_auth);
         }
     }
 
     /**
+     * The current session will be authenticated using this token_auth.
+     * It will overwrite the previous Auth object.
+     *
+     * @param string $tokenAuth
+     * @return void
+     */
+    private static function doReloadAuthUsingTokenAuth($tokenAuth)
+    {
+        /**
+         * Triggered when authenticating an API request, but only if the **token_auth**
+         * query parameter is found in the request.
+         *
+         * Plugins that provide authentication capabilities should subscribe to this event
+         * and make sure the global authentication object (the object returned by `StaticContainer::get('Piwik\Auth')`)
+         * is setup to use `$token_auth` when its `authenticate()` method is executed.
+         *
+         * @param string $token_auth The value of the **token_auth** query parameter.
+         */
+        Piwik::postEvent('API.Request.authenticate', array($tokenAuth));
+        Access::getInstance()->reloadAccess();
+        SettingsServer::raiseMemoryLimitIfNecessary();
+    }
```

Comment:
I can rename it to `doReallyReloadAuthUsingTokenAuth`. That method is called by `reloadAuthUsingTokenAuth` but also when restoring the token. Here we definitely want to reload auth even `shouldReloadAuthUsingTokenAuth` would return `false`. That's why I added that method.